core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -88,140 +53,388 @@ class Queue[A] private (
    * It places the values in the queue but if there is no room it will not enqueue them and returns false
    *
    */
-  final def offerAll(as: Iterable[A]): UIO[Boolean] =
-    for {
-      _ <- checkShutdownState
-
-      remaining <- IO.effectTotalWith { platform =>
-                    val pTakers                = if (queue.isEmpty()) unsafePollN(takers, as.size) else List.empty
-                    val (forTakers, remaining) = as.splitAt(pTakers.size)
-                    (pTakers zip forTakers).foreach {
-                      case (taker, item) => unsafeCompletePromise(taker, item, platform)
-                    }
-                    remaining
-                  }
-
-      added <- if (remaining.nonEmpty) {
-                // not enough takers, offer to the queue
-                for {
-                  surplus <- IO.effectTotalWith { platform =>
-                              val as = unsafeOfferAll(queue, remaining.toList)
-                              unsafeCompleteTakers(platform)
-                              as
-                            }
-                  res <- if (surplus.isEmpty) IO.succeed(true)
-                        else
-                          strategy.handleSurplus(surplus, queue) <* IO.effectTotalWith(
-                            platform => unsafeCompleteTakers(platform)
-                          )
-                } yield res
-              } else IO.succeed(true)
-    } yield added
+  def offerAll(as: Iterable[A]): ZIO[R, E, Boolean]
 
   /**
    * Waits until the queue is shutdown.
    * The `IO` returned by this method will not resume until the queue has been shutdown.
    * If the queue is already shutdown, the `IO` will resume right away.
    */
-  final val awaitShutdown: UIO[Unit] =
-    for {
-      p  <- Promise.make[Nothing, Unit]
-      io = p.succeed(()).void
-      _ <- IO.flatten(shutdownHook.modify {
-            case None       => (io, None)
-            case Some(hook) => (IO.unit, Some(hook *> io))
-          })
-      _ <- p.await
-    } yield ()
+  def awaitShutdown: ZIO[R, E, Unit]
 
   /**
    * Retrieves the size of the queue, which is equal to the number of elements
    * in the queue. This may be negative if fibers are suspended waiting for
    * elements to be added to the queue.
    */
-  final val size: UIO[Int] = checkShutdownState.map(_ => queue.size - takers.size + strategy.surplusSize)
+  def size: ZIO[R, E, Int]
```

Comment:
I think this one could still be a `UIO[Int]`. The `mapM` and `contramapM` cases don't change it.

Reply:
Good point. I had initially thought that other combinators may modify the behavior and introduce potential failure, but it's probably a better design to keep this, `awaitShutdown` and `shutdown` as `UIO`. Will change.